### PR TITLE
Fix a panic in the `parquet_encode` processor

### LINKED ISSUE #2885: Panic when encoding same batch as parquet across multiple outputs
If you `fan_out` the same messages to multiple outputs that each try to encode with a `parquet_encode` processor, you get a panic.

```
processor_resources:
  - label: parquet_encoder
    parquet_encode:
      default_compression: zstd
      schema:
        - name: document_id
          type: UTF8

...

output:
  broker:
    outputs:
      - aws_s3:
          batching:
            processors:
              - resource: parquet_encoder
      - aws_s3:
          batching:
            processors:
              - resource: parquet_encoder
```

Panic usually looks like this, but I've seen some other variations:
```
fatal error: concurrent map writes

goroutine 119 [running]:
github.com/redpanda-data/connect/v4/internal/impl/parquet.scrubJSONNumbersObj(...)
        /go/pkg/mod/github.com/redpanda-data/connect/v4@v4.35.3/internal/impl/parquet/util.go:41
github.com/redpanda-data/connect/v4/internal/impl/parquet.scrubJSONNumbers({0x4b92440?, 0xc00232e300})
        /go/pkg/mod/github.com/redpanda-data/connect/v4@v4.35.3/internal/impl/parquet/util.go:30 +0x1ed
github.com/redpanda-data/connect/v4/internal/impl/parquet.(*parquetEncodeProcessor).ProcessBatch(0xc0019f4fa0, {0x3e8?, 0x822a6f?}, {0xc01d13a008, 0x3e8, 0x60?})
        /go/pkg/mod/github.com/redpanda-data/connect/v4@v4.35.3/internal/impl/parquet/processor_encode.go:281 +0x1b8
github.com/redpanda-data/benthos/v4/public/service.(*airGapBatchProcessor).ProcessBatch(0xc001c1cdf0, 0xc018ef0c60, {0xc01afa8008, 0x3e8, 0x3e8})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/public/service/processor.go:112 +0xf2
github.com/redpanda-data/benthos/v4/internal/component/processor.(*v2BatchedToV1Processor).ProcessBatch(0xc00319a6c0, {0x61a6f10, 0xc001c08370}, {0xc01afa8008, 0x3e8, 0x3e8})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/component/processor/auto_observed.go:223 +0x208
github.com/redpanda-data/benthos/v4/internal/impl/pure.(*resourceProcessor).ProcessBatch.func1({0x6170058?, 0xc00319a6c0?})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/impl/pure/processor_resource.go:84 +0x4f
github.com/redpanda-data/benthos/v4/internal/manager.(*Type).AccessProcessor.func1({0x6170058?, 0xc00319a6c0?})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/manager/type.go:674 +0x6e
github.com/redpanda-data/benthos/v4/internal/manager.(*liveResource[...]).RAccess(0xc001f792f0?, 0xc001b1fb60?)
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/manager/live_resources.go:41 +0x98
github.com/redpanda-data/benthos/v4/internal/manager.(*liveResources[...]).RAccess(0x621e7e0, {0xc001b1fb60, 0xf}, 0xc00e019b08)
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/manager/live_resources.go:155 +0xc6
github.com/redpanda-data/benthos/v4/internal/manager.(*Type).AccessProcessor(0x4eaf200?, {0x458601?, 0x18?}, {0xc001b1fb60?, 0x418798?}, 0x1?)
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/manager/type.go:669 +0x75
github.com/redpanda-data/benthos/v4/internal/impl/pure.(*resourceProcessor).ProcessBatch(0xc0023a3830, {0x61a6f10, 0xc001c08370}, {0xc01afa8008, 0x3e8, 0x3e8})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/impl/pure/processor_resource.go:83 +0x110
github.com/redpanda-data/benthos/v4/internal/component/processor.ExecuteAll({0x61a6f10, 0xc001c08370}, {0xc001e8ba80, 0x2, 0x412245?}, {0xc001f55ca8?, 0x1, 0xc00167b070?})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/component/processor/execute.go:19 +0x1c7
github.com/redpanda-data/benthos/v4/internal/batch/policy.(*Batcher).flushAny(0xc000b57110, {0x61a6f10, 0xc001c08370})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/batch/policy/policy.go:167 +0x1dd
github.com/redpanda-data/benthos/v4/internal/batch/policy.(*Batcher).Flush(0xc0190e6238?, {0x61a6f10?, 0xc001c08370?})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/batch/policy/policy.go:134 +0x30
github.com/redpanda-data/benthos/v4/internal/component/output/batcher.(*Impl).loop(0xc001c46910)
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/component/output/batcher/batcher.go:128 +0x607
created by github.com/redpanda-data/benthos/v4/internal/component/output/batcher.(*Impl).Consume in goroutine 1
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/component/output/batcher/batcher.go:178 +0xb6
```

My guess is that the processor isn't thread-safe and is being shared when used as a resource. I assumed the resource would construct an independent, duplicate, processor for each output.

### PR DESCRIPTION
**Pull request by mihaitodor**
Fixes #2885.

### CODE REVIEW

**Review by rockwotj — APPROVED** [review]
LGTM

File: internal/impl/parquet/processor_encode.go [hunk 1]
```diff
@@ -270,6 +270,7 @@ func (s *parquetEncodeProcessor) ProcessBatch(ctx context.Context, batch service
 	buf := bytes.NewBuffer(nil)
 	pWtr := parquet.NewGenericWriter[any](buf, s.schema, parquet.Compression(s.compressionType))
 
+	batch = batch.Copy()
 	rows := make([]any, len(batch))
 	for i, m := range batch {
 		ms, err := m.AsStructured()
```
Comment (Jeffail) [inline internal/impl/parquet/processor_encode.go:276]: AsStructured
```suggestion
		ms, err := m.AsStructuredMut()
```

Comment (mihaitodor) [reply internal/impl/parquet/processor_encode.go]: Oh, OK. Guess the batch still needs to be copied as well since the test I added fails otherwise.

Comment (mihaitodor) [reply internal/impl/parquet/processor_encode.go]: Also, I had a quick look through this API and found this redundant assignment: https://github.com/redpanda-data/benthos/blob/50a5cd25a0c1dcef357a1467a0f2ddf556bee8d7/internal/message/data.go#L74 I wonder why the linter isn't catching it. Might be worth revising that as well.

**Review by Jeffail — APPROVED** [review]
LGTM, thanks @mihaitodor!

File: internal/impl/parquet/processor_encode.go (hunk 1 above)
Comment (Jeffail) [reply internal/impl/parquet/processor_encode.go]: Wow good catch, I'll raise a PR for it.

**Pull request merged**